File: backend/services/atomic/task/controller/RecurrenceController.js
const Recurrence = require("../model/Recurrence");

module.exports = {
  async getRecurrenceById(req, res) {
    try {
      const recurrence = await Recurrence.getById(req.params.id);
      res.status(200).json(recurrence);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  async getRecurrenceByTaskId(req, res) {
    try {
      const recurrences = await Recurrence.getByTaskId(req.params.taskId);
      res.status(200).json(recurrences);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  async createRecurrence(req, res) {
    try {
      console.log("Payload received in atomic service (createRecurrence):", req.body);
      const recurrence = new Recurrence(req.body);
      await recurrence.create();
      res.status(201).json(recurrence);
    } catch (error) {
      console.error("Error in atomic service (createRecurrence):", error.message);
      res.status(500).json({ error: error.message });
    }
  },

  async updateRecurrence(req, res) {
    try {
      console.log("Recurrence ID:", req.params.id);
      console.log("Request Body:", req.body);

      const recurrence = new Recurrence({ id: req.params.id, ...req.body });
      await recurrence.update();
      res.status(200).json(recurrence);
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },

  async deleteRecurrence(req, res) {
    try {
      const recurrence = new Recurrence({ id: req.params.id });
      await recurrence.delete();
      res.status(200).json({ message: "Recurrence deleted successfully" });
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  },
};

File: backend/services/composite/manage-task/src/main/java/com/spm/manage_task/controller/RecurrenceController.java
package com.spm.manage_task.controller;

import java.util.List;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.spm.manage_task.dto.RecurrenceDto;
import com.spm.manage_task.services.RecurrenceService;

@RestController
@RequestMapping("api/recurrence")
public class RecurrenceController {

    private final RecurrenceService recurrenceService;

    public RecurrenceController(RecurrenceService recurrenceService) {
        this.recurrenceService = recurrenceService;
    }

    // GET a specific recurrence by ID
    @GetMapping("/{recurrenceId}")
    public ResponseEntity<RecurrenceDto> getRecurrenceById(@PathVariable String recurrenceId) {
        RecurrenceDto recurrence = recurrenceService.getRecurrenceById(recurrenceId);
        return ResponseEntity.ok(recurrence);
    }

    // GET all recurrences for a specific task
    @GetMapping("/task/{taskId}")
    public ResponseEntity<List<RecurrenceDto>> getRecurrencesByTaskId(@PathVariable String taskId) {
        List<RecurrenceDto> recurrences = recurrenceService.getRecurrencesByTaskId(taskId);
        return ResponseEntity.ok(recurrences);
    }

    // POST to create a new recurrence
    @PostMapping("/")
    public ResponseEntity<String> createRecurrence(@RequestBody RecurrenceDto recurrenceDto) {
        System.out.println("Payload received in composite controller (createRecurrence): " + recurrenceDto);
        recurrenceService.createRecurrence(recurrenceDto);
        return ResponseEntity.status(201).body("Recurrence created successfully");
    }

    // PUT to update an existing recurrence
    @PutMapping("/{recurrenceId}")
    public ResponseEntity<String> updateRecurrence(@PathVariable String recurrenceId, @RequestBody RecurrenceDto recurrenceDto) {
        recurrenceService.updateRecurrence(recurrenceId, recurrenceDto);
        return ResponseEntity.ok("Recurrence updated successfully");
    }

    // DELETE a recurrence by ID
    @DeleteMapping("/{recurrenceId}")
    public ResponseEntity<String> deleteRecurrence(@PathVariable String recurrenceId) {
        recurrenceService.deleteRecurrence(recurrenceId);
        return ResponseEntity.ok("Recurrence deleted successfully");
    }
}